ddtrace/ext/ci.py

```diff
@@ -41,28 +43,33 @@
 
 
 def _normalize_ref(name):
+    # type: (Optional[str]) -> Optional[str]
     return _RE_TAGS.sub("", _RE_ORIGIN.sub("", _RE_REFS.sub("", name))) if name is not None else None
 
 
 def _filter_sensitive_info(url):
+    # type: (Optional[str]) -> Optional[str]
     return _RE_URL.sub("\\1", url) if url is not None else None
 
 
 def tags(env=None):
+    # type: (Optional[MutableMapping[str, str]]) -> MutableMapping[str, str]
+    """Extract and set tags from provider environ."""
     env = os.environ if env is None else env
-    tags = {}
+    tags = {}  # type: MutableMapping[str, Optional[str]]
     for key, extract in PROVIDERS:
         if key in env:
             tags = extract(env)
             break
 
+    git_info = git.extract_git_metadata(tags)
```

Comment:
This will raise an exception if a git executable is not found. Is this okay? Do we always assume git is available?

A lot of stuff here is git-centric but this would mean that the CI app won't work at all with other SCMs

Reply:
This is a great point! In case the git executable isn't found and we get an error from the subprocess command, the extracted metadata defaults to empty strings. For the CI app to work, they need 1) branch name, 2) repo url, and 3) commit SHA which are normally available in the CI providers' environment variable - so if no extra git information is provided, we'll still have basic CI app functioning.

From the CI app team, we currently don't support other SCMs at the moment and are assuming that all CI usage is via git.

Reply:
If a git executable isn't found then I think you get a `FileNotFoundError`:

```python
>>> subprocess.Popen(["git2", "status"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kyle.verhoog/.pyenv/versions/3.6.10/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/Users/kyle.verhoog/.pyenv/versions/3.6.10/lib/python3.6/subprocess.py", line 1364, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'git2': 'git2'
```

Reply:
(To add to what @Kyle-Verhoog said, `FileNotFoundError` is Py3 only. You'll get a good old `OSError` on Py2. 😢 )